iceoryx2-bb/cxx/include/iox2/bb/semantic_string.hpp

```diff
@@ -0,0 +1,460 @@
+// Copyright (c) 2023 by Apex.AI Inc. All rights reserved.
+// Copyright (c) 2025 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+#ifndef IOX2_BB_SEMANTIC_STRING_HPP
+#define IOX2_BB_SEMANTIC_STRING_HPP
+
+#include "iox2/legacy/expected.hpp"
+#include "iox2/legacy/logging.hpp"
+#include "iox2/legacy/string.hpp"
+
+#include <cstdint>
+
+namespace iox2 {
+namespace bb {
+/// @brief Defines errors which can occur when modifying or creating a SemanticString
+enum class SemanticStringError : uint8_t {
+    InvalidContent,
+    ExceedsMaximumLength
+};
+
+template <uint64_t Capacity>
+using DoesContainInvalidCharacter = bool (*)(const legacy::string<Capacity>& value);
+
+template <uint64_t Capacity>
+using DoesContainInvalidContent = bool (*)(const legacy::string<Capacity>& value);
+
+/// @brief The SemanticString is a string which has an inner syntax and restrictions
+///         to valid content. Examples are for instance
+///         * FileNames, no slashes etc.
+///        The SemanticString shall be a basic building block to create string
+///        types with a semantic contract. It is used via inheritance. The user
+///        has to also define the maximum capacity, a callable which defines
+///        invalid characters as well as a callable which defines invalid content.
+/// @code
+/// bool user_name_does_contain_invalid_characters(const string<platform::MAX_USER_NAME_LENGTH>& value) noexcept;
+/// bool user_name_does_contain_invalid_content(const string<platform::MAX_USER_NAME_LENGTH>& value) noexcept;
```

Comment:
Why are these functions named `user_name_*` but the class was renamed to `FileName`?